chemprop/data/datasets.py

```diff
@@ -252,11 +252,11 @@ def normalize_inputs(
             case "X_d":
                 X = None if np.all(self.X_d == None) else self.X_d
             case "V_f":
-                X = None if np.all(self.V_fs == None) else np.concatenate(self.V_fs, axis=0)
+                X = None if all(V_f is None for V_f in self.V_fs) else np.concatenate(self.V_fs, axis=0)
             case "E_f":
-                X = None if np.all(self.E_fs == None) else np.concatenate(self.E_fs, axis=0)
+                X = None if all(E_f is None for E_f in self.E_fs) else np.concatenate(self.E_fs, axis=0)
             case "V_d":
-                X = None if np.all(self.V_ds == None) else np.concatenate(self.V_ds, axis=0)
+                X = None if all(V_d is None for V_d in self.V_ds) else np.concatenate(self.V_ds, axis=0)
```

Comment:
For all of these, we could check if the corresponding dimension is equal to 0 e.g. `X = None if self.d_vd == 0 else ...` Those dimension properties also need to check if X_d, etc. are None. Using those properties would reduce duplicate code.

That said, I don't think those dimension properties have the right logic. For each of X_d, V_f, E_f, and V_d, the feature/descriptor needs to be given for all datapoints in the dataset if it is given to one datapoint. This means to get the dimension of d_vd, we only need to look at the first datapoint's column length, which is what it currently does (`self.V_ds[0].shape[1]`). It also means we only need to check if the first datapoint has V_d as None (`self.V_ds[0] == None`) instead of checking all of them.

Reply:
Thanks for the suggestion! I think for `X = None if self.d_vd == 0 else ...` to work we need to add `self.d_vd` to reaction dataset and set them to 0. I'm ok with that and I think it can help us to have cleaner code.